x8.md

```diff
@@ -203,230 +76,18 @@ The following messages types MUST be supported:
 * `user-message` is a private message
 * `group-user-message` is a message to the private group.
 
-## Clock vs Timestamp and message ordering
+# Clock vs Timestamp and message ordering
 
 `timestamp` MUST be Unix time calculated when the message is created. Because the peers in the Whisper network should have synchronized time, `timestamp` values should be fairly accurate among all Whisper network participants.
 
 `clock` SHOULD be calculated using the algorithm of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps). When there are messages available in a chat, `clock`'s value is calculated based on the last received message in a particular chat: `last-message-clock-value + 1`. If there are no messages, `clock` is initialized with `timestamp`'s value.
 
 `clock` value is used for the message ordering. Due to the used algorithm and distributed nature of the system, we achieve casual ordering which might produce counterintuitive results in some edge cases. For example, when one joins a public chat and sends a message before receiving the exist messages, their message `clock` value might be lower and the message will end up in the past when the historical messages are fetched.
```

Comment:
What happens if clock or timestamp is deliberately set incorrectly?

(I understand a lot of these questions might not be able to answer straight away, so can just add as TODOs or so)

Reply:
If `clock` is deliberately set incorrectly your message can be  orddered arbitrarely (in the past or in the future), and that's unavoidable with logical clocks. To prevent this we discard messages that are "too" far off in the future, while we are ok on messages in the past (as they might come from a mailserver).

Timestamps currently is used as well for ordering (grouping by day), but that's a regression introduced during some performance optimization, which might result in messages not being casually ordered if the sender has it's clock drifted for more than 24 hours. Otherwise it is only used to show a timestamp next to each message.

Reply:
Yeah makes sense, so from a protocol pov:

- note on it isn't cryptographically ordering events and this can be abused by bad actor
-  note on not relaying on timestamps

Then hints for app developers to:
- consider discarding messages too far in future (clock)
- possibly using timestamp for ordering/grouping per day with real clock (~)

Reply:
Not sure, it makes sense to add what is not (it is not many things, it is only a few), and for the technical reader will know that lamport timestamps are just that, it's not a blockchain or a consensus based mechanism.

Also "cryptographically ordering events" is not a well defined term by any means, you might use cryptography to preserve ordering, but what is in your mind "cryptographical order"?

Reply:
It is a simple attack model and it weakens ability to order events in face of basic adversary. It might not matter a lot for chat,  but it definitely could for other applications like state channels etc.

Cryptographic hash function guarantees means you can maintain certain invariants which you can’t under Byzantine model (it is easy to come up with predictable integers)

Reply:
It’s an instance of don’t trust user input, especially in p2p

Reply:
I don't disagree we should add the fact that currently anyone can insert messages in the conversation, it's just a matter of how.

I would simply state the fact that currently we discard messages that have an higher clock value then `timestamp` + 1 month, while we accept any message that has a lower clock value than `timestamp` (this is the same as briar I believe, though briar uses a day).

The rest I don't think it belongs here, again, we should describe what it is and not what is not, and explain what are the attack vectors.

If we want to add something on the lines you proposed, it needs to be well defined at least:
i.e:
"cryptographical order" gives 4 results on google so I doubt people will understand what you mean ( I don't, best guess is that is the parent id stuff, or you are thinking about a blockchain like consensus model)

> Cryptographic hash function guarantees means you can maintain certain invariants 

list which invariants you  think are useful and to solve which problem, in our use case

> It’s an instance of don’t trust user input, especially in p2p

again, this kind of comments are not really useful, either you take the time to deal with the complexity and nuances, or they have the same usefulness as "don't break compatibility"

Reply:
> cryptographical order

This was sloppily written and is a shorthand for ordering messages like a DAG, based on message ids as dependencies that are cryptographic hashes. It proves that you have seen at least that message and can't arbitrarily claim you are from the future. Sorry if this wasn't clear. If it still isn't I'll add it in a separate PR next week.

> deal with the complexity and nuances, or they have the same usefulness as "don't break compatibility"

It's really not that complex and also not that big of a deal, it's just a caveat. It's important for people writing clients to be aware that they don't rely on it. I disagree that general principles like "don't trust user input" and "don't break compatibility" aren't useful - I guess it depends on having a common understanding of these being goals and then working together to solve them. In this specific instance it's fairly straightforward as an enhancement (that may or may not be needed, depending on the application), whereas in other cases (compatibility and whisper topic partitioning) it might be more difficult and require more careful thinking, as you rightly point out. To me, it's useful to re-iterate (presumably shared) goals to remind ourselves of them and think about how they can be achieved, or what the limitations are.

Reply:
> It proves that you have seen at least that message and can't arbitrarily claim you are from the future.

This approach though brings other sort of issues, and it should fit our use case without changing the way things work now (should work on public chats, you should be able to post without having seen any message etc):

- difficult to order with partial information, i.e you have X1 -> X2 -> X4 -> X6 -> X8  X6 and X4 can't be ordered with a single parent-id, you can have multiple parent id but you are just making this less likely to happen

- you can still insert messages in history, unless you implement some form of consensus

- you still need a meaningful way to break ties (multiple messages with the same parent id)

- can't post if you haven't fetched a message without some other mean of ordering etc

so again I would not add references to something that is not specified, unless there's a know implementation that people can reference ( which I am not aware of, closest is briar but only uses that only for threads).

Also having a higher/predictable clock value is not a big issue in our case, (you want your message to have the higher timestamp, other clients will do +1), we cap it just to allow messages from users who did not see any message to be ordered meaningfully (causality it is preserved even without this though, a parent-id method would likely need similar arrangement if we want to consider such cases).

>  I disagree that general principles like "don't trust user input" and "don't break compatibility" aren't useful

I might have worded this too strongly, I just feel that they provide little information, we all have enough working experience to know that we should not be blindly trusting user input, and we don't as a matter of fact, although we could be better. 
I would again just mention the tradeoffs (messages can be inserted in the past, high clock value case is already mentioned), and we can think about how to address those in a separate place.

Reply:
> `clock` value is used for the message ordering.

will be

> `clock` value SHOULD is used for the message ordering.

Additionally, I will:
- add info about messages being inserted in the past,
- suggest that `timestamp` is recommended in the presentation layer,
- that this mechanism can be easily abused and clients MAY chose to drop messages too old or from the future.

Reply:
sounds good

> clock value SHOULD be used for the message ordering.

typo

`timestamp` is tricky, I personally think we should remove it, use whisper timestamps for the presentation layer (not sorting, not even grouping), and we could check against whisper timestamp/clock-value difference to handle some of the most obvious abuses

Reply:
That's a good point. I will add that for the presentation layer, Whisper envelope time should be used. In a separate PR (for better visibility and clean process of decision making), we can deprecate `timestamp` if we think it's not needed.